src/cpp/ext/otel/otel_server_call_tracer.h

```diff
@@ -21,28 +21,130 @@
+            size_t index) {
+          auto* labels_injector = plugin_option.labels_injector();
+          if (labels_injector != nullptr) {
+            labels_injector->AddLabels(
+                send_initial_metadata,
+                injected_labels_from_plugin_options_[index].get());
+          }
+          return true;
+        },
+        otel_plugin_);
+  }
+
+  void RecordSendTrailingMetadata(
+      grpc_metadata_batch* /*send_trailing_metadata*/) override;
 
-  grpc_core::ServerCallTracer* CreateNewServerCallTracer(
-      grpc_core::Arena* arena,
-      const grpc_core::ChannelArgs& channel_args) override;
+  void RecordSendMessage(const grpc_core::SliceBuffer& send_message) override {
+    RecordAnnotation(
+        absl::StrFormat("Send message: %ld bytes", send_message.Length()));
+  }
+  void RecordSendCompressedMessage(
+      const grpc_core::SliceBuffer& send_compressed_message) override {
+    RecordAnnotation(absl::StrFormat("Send compressed message: %ld bytes",
+                                     send_compressed_message.Length()));
+  }
 
-  bool IsServerTraced(const grpc_core::ChannelArgs& args) override;
+  void RecordReceivedInitialMetadata(
+      grpc_metadata_batch* recv_initial_metadata) override;
+
+  void RecordReceivedMessage(
+      const grpc_core::SliceBuffer& recv_message) override {
+    RecordAnnotation(
+        absl::StrFormat("Received message: %ld bytes", recv_message.Length()));
+  }
+  void RecordReceivedDecompressedMessage(
+      const grpc_core::SliceBuffer& recv_decompressed_message) override {
+    RecordAnnotation(absl::StrFormat("Received decompressed message: %ld bytes",
+                                     recv_decompressed_message.Length()));
+  }
+
+  void RecordReceivedTrailingMetadata(
+      grpc_metadata_batch* /*recv_trailing_metadata*/) override {}
+
+  void RecordCancel(grpc_error_handle /*cancel_error*/) override {
+    elapsed_time_ = absl::Now() - start_time_;
+  }
+
+  void RecordEnd(const grpc_call_final_info* final_info) override;
+
+  void RecordAnnotation(absl::string_view /*annotation*/) override {
+    // Not implemented
+  }
+
+  void RecordAnnotation(const Annotation& /*annotation*/) override {
+    // Not implemented
+  }
+  std::shared_ptr<grpc_core::TcpTracerInterface> StartNewTcpTrace() override {
+    // No TCP trace.
+    return nullptr;
+  }
 
  private:
+  absl::string_view MethodForStats() const {
+    absl::string_view method = absl::StripPrefix(path_.as_string_view(), "/");
+    if (registered_method_ ||
+        (otel_plugin_->generic_method_attribute_filter() != nullptr &&
+         otel_plugin_->generic_method_attribute_filter()(method))) {
+      return method;
+    }
+    return "other";
+  }
+
+  absl::Time start_time_;
+  absl::Duration elapsed_time_;
+  grpc_core::Slice path_;
+  bool registered_method_;
+  // TODO(yashykt): It's wasteful to do this per call. When we re-haul the stats
```

Comment:
We should probably take care of this TODO as part of this PR.  I think we can put this in the `ScopeConfig` object.  Note that we can have separate `ScopeConfig` subclasses for client-side and server-side, so that we don't waste memory on fields that we need on only one side.

Reply:
I think I may have steered you wrong here.  When I read this TODO, I thought this was something that we were initializing per-call even though it didn't depend on per-call data, which is why I suggested initializing it per-server-connection instead.  But looking at the code, it looks like the value actually depends on the call's initial metadata.  So we're now storing a single copy for the connection but setting the data on a per-call basis, which isn't thread-safe and will cause conflicts between calls.

@yashykt should weigh in here about what this TODO was supposed to be about (maybe it was about moving to a metadata lookup service?), but unless he has a better idea, I think this needs to go back to being per-call.

Reply:
I think I added the TODO at the wrong line. This is about the `active_plugin_options_view_` which only needs to be calculated once per server. This can be removed

Reply:
Done. Reverted.